libraries/AP_Networking/AP_Networking_CAN.cpp

```diff
@@ -0,0 +1,226 @@
+{
+#ifndef HAL_BOOTLOADER_BUILD
+    while (!hal.scheduler->is_system_initialized()) {
+        hal.scheduler->delay(100);
+    }
+#endif
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "CAN_MCAST: starting");
+
+    ObjectBuffer<AP_HAL::CANFrame> *frame_buffers[HAL_NUM_CAN_IFACES] {};
+
+    for (uint8_t bus=0; bus<HAL_NUM_CAN_IFACES; bus++) {
+        auto *cbus = get_caniface(bus);
+        if (cbus == nullptr) {
+            continue;
+        }
+        if (bus_mask & (1U<<bus)) {
+            mcast_sockets[bus] = NEW_NOTHROW SocketAPM(true);
+            if (mcast_sockets[bus] == nullptr) {
+                GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "CAN_MCAST[%u]: failed to create socket", unsigned(bus));
+                continue;
+            }
+        }
+        char address[] = MCAST_ADDRESS_BASE;
+        const uint32_t buffer_size = 20; // good for fw upload
+        uint8_t callback_id = 0;
+
+        address[strlen(address)-1] = '0' + bus;
+        if (!mcast_sockets[bus]->connect(address, MCAST_PORT)) {
+            GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "CAN_MCAST[%u]: failed to connect", unsigned(bus));
+            goto de_allocate;
+        }
+
+        if (!cbus->register_frame_callback(
+                FUNCTOR_BIND_MEMBER(&AP_Networking_CAN::can_frame_callback, void, uint8_t, const AP_HAL::CANFrame &),
+                callback_id)) {
+            GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "CAN_MCAST[%u]: failed to register", unsigned(bus));
+            goto de_allocate;
+        }
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+        // tell the ethernet interface that we want to receive all
+        // multicast packets
+        ETH->MACPFR |= ETH_MACPFR_PM;
+#endif
+
+        frame_buffers[bus] = NEW_NOTHROW ObjectBuffer<AP_HAL::CANFrame>(buffer_size);
+        if (frame_buffers[bus] == nullptr) {
+            GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "CAN_MCAST[%u]: failed to allocate buffers", unsigned(bus));
+            goto de_allocate;
+        }
+        continue;
+
+    de_allocate:
+        delete mcast_sockets[bus];
+        mcast_sockets[bus] = nullptr;
+    }
+
+
+    // main loop
+    while (true) {
+        const uint32_t delay_us = 100; // limit to 10k packets/s
+#ifndef HAL_BOOTLOADER_BUILD
+        hal.scheduler->delay_microseconds(delay_us);
+#else
+        thread_sleep_us(delay_us);
+#endif
+        for (uint8_t bus=0; bus<HAL_NUM_CAN_IFACES; bus++) {
+            if (mcast_sockets[bus] == nullptr) {
+                continue;
+            }
+
+            struct mcast_pkt pkt;
+            const ssize_t ret = mcast_sockets[bus]->recv((void*)&pkt, sizeof(pkt), 0);
+            if (ret > MCAST_HDR_LENGTH && ret <= sizeof(pkt)) {
+                const uint8_t data_len = ret - MCAST_HDR_LENGTH;
+                bool is_canfd = false;
+#if HAL_CANFD_SUPPORTED
+                is_canfd = (pkt.flags & MCAST_FLAG_CANFD) != 0;
+#endif
+                if (pkt.magic != MCAST_MAGIC) {
+                    continue;
```

Comment:
Is magic intended as a compatibilty flag? If so, perhaps it's worth a one-time warning to the user that an incompatible CAN mcast packet was received. Otherwise, wouldn't this silently drop the data? It could be hard to figure out why two devices with different magic values aren't going to communicate.

Reply:
no, it is to ensure that some other multicast user that happens to choose the same port doesn't inject bad data. I tried to reserve a port with IANA but they rejected my application

Reply:
Maybe that deserves a comment?